### PR TITLE
knot-resolver: 4.3.0 -> 5.0.0 -> 5.0.1

### PR DESCRIPTION
**Pull request by vcunat**
###### Motivation for this change

https://gitlab.labs.nic.cz/knot/knot-resolver/tags/v5.0.0

The service needed lots of changes.  A few smaller changes are added into the PR, e.g. replacement for PR #72014.  See the commit messages for details.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] N/A Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] N/A Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after):
   negligible increase (9576 bytes)
- [ ] N/A Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
- [x] _Tested the NixOS service with various combinations of options._

### CODE REVIEW

File: nixos/modules/services/networking/kresd.nix [hunk 1]
```diff
@@ -93,73 +126,20 @@ in {
...
+    systemd.targets.kresd = {
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "kres-cache-gc.service" ]
+        ++ map (i: "kresd@${toString i}.service") (range 1 cfg.instances);
     };
+    systemd.services."kresd@".serviceConfig.ExecStart = "${package}/bin/kresd --noninteractive "
+      + "-c ${package}/lib/knot-resolver/distro-preconfig.lua -c ${configFile}";
+
+    # TODO: try somehow utilizing ${package}/lib/tmpfiles.d/*
+    systemd.tmpfiles.rules = [
```
Comment (Mic92) [inline nixos/modules/services/networking/kresd.nix]: Right now you can do https://github.com/NixOS/nixpkgs/pull/78426/files instead to use it directly.

Comment (vcunat) [reply nixos/modules/services/networking/kresd.nix]: Thanks, pushed.

**Comment by vcunat** [comment]
Note: upstream and packages on other distros now use the "knot-resolver" name on more places, e.g. `/var/cache/knot-resolver/` instead of `/var/cache/kresd/`, user/group name and `/etc/knot-resolver/`.  The cost/benefit ratio of doing such changes didn't seem very favorable to me, so I didn't touch those &ndash; at least for now.  _Well, moving `/etc/` stuff is easy on NixOS, but users can't edit it anyway and they might not have motivation to even read from there, so it's perhaps unimportant._

**Comment by Mic92** [comment]
I would migrate to upstream defaults soonish as it makes it easier to follow upstream documentation.

**Comment by vcunat** [comment]
- users+group: drop the current IDs?  And either assign now ones or migrate to that "dynamic" scheme of systemd?
- cache directory: the old one will just be left rotting, I guess?  I consider adding tmpfiles spec that removes the old one.

**Comment by vcunat** [comment]
I wonder whether to put all the changes into a single PR, but it probably doesn't really matter, given the (expected) number of reviewers.

**Comment by Mic92** [comment]
>     * user+group: drop the current IDs?  And either assign now ones or migrate to that "dynamic" scheme of systemd?  (I'm not aware of a good way of renaming them, but I haven't looked much.)

knot-resolver looks like a good candidate for `DynamicUser` except that users might need a user/group in case certificates are generated via the acme module
for chowning them (or any other secret managed by nixops/krops). It is not possible right now to rename uid/gids, however old ids just get recycled eventually.

> 
>     * cache directory: the old one will just be left rotting, I guess?  I consider instead adding tmpfiles spec that removes the old one.

I fear that deleting user data causes more harm than it helps. I would just leave it as it is.

**Comment by vcunat** [comment]
Right, if it would make it more difficult to give kresd read access to ACME certs, I don't think we want to go that way. So I'll get new fixed IDs, I suppose.


There should be no "user data" in the cache directory.  I know we don't precisely follow FHS or similar, but still.
> `/var/cache/` Flushing this directory should have no effect on operation of programs, except for increased runtimes necessary to rebuild these caches.

Still, if there are any doubts, not touching the old cache is the easiest approach.

**Comment by Mic92** [comment]
> Right, if it would make it more difficult to give kresd read access to ACME certs, I don't think we want to go that way. So I'll get new fixed IDs, I suppose.
> 

The IDs don't need to be fixed. They can also be allocated at activation time, however a user/group name should be allocated for `chown()` to work.

> There should be no "user data" in the cache directory. I know we don't precisely follow FHS or similar, but still.
> 
> > `/var/cache/` Flushing this directory should have no effect on operation of programs, except for increased runtimes necessary to rebuild these caches.
> 
> Still, if there are any doubts, not touching the old cache is the easiest approach.

**Comment by vcunat** [comment]
> The IDs don't need to be fixed. [...]

Whenever the IDs change, the file ownership won't match anymore.  I think ownership can be force-changed by switching to "e" rule in tmpfiles, but overall I wonder about cost/benefit ratio.

**Comment by Mic92** [comment]
> > The IDs don't need to be fixed. [...]
> 
> Whenever the IDs change, the file ownership won't match anymore. I think ownership can be force-changed by switching to "e" rule in tmpfiles, but overall I wonder about cost/benefit ratio.

If `CacheDirectory` is used than systemd should take care of fixing the ownership.

**Comment by vcunat** [comment]
Right, that's an option, if we remove the possibility to configure `cacheDir` location (or restrict it being within `/var/cache/`).

**Comment by Mic92** [comment]
I don't think this option is often used anyway. Deprecating it makes things easier and it is always possible to put a different filesystem using a bind mount. Also see: 5c18c08da9fd13d583b56d3b8cc698621ce8cd16

**Comment by vcunat** [comment]
OK, let me remove that option.

By the way, larger deployments want to have cache in tmpfs to reduce the I/O load (and persistence across reboots isn't a big deal), and we/upstream recommend to mount a separate tmpfs on the cache directory instead of relocating it.  Perhaps I'll add a simple NixOS option for this later, though I haven't yet looked into whether to do this via systemd features or plain `filesystems."/var/cache/knot-resolver"`.

**Comment by vcunat** [comment]
I incorporated all that's been discussed now, I think.  Also minor fixes/improvements were added, and the git history was cleaned up.  I'm using the service already, as usual.

All the namings (users, paths) should be the same now as in other distributions and (upstream) docs.  I kept the `services.kresd` name, as I'm not aware of any equivalent elsewhere anyway &ndash; but it would be cheap to change (compatible).

**Review by Mic92 — APPROVED** [review]
Looks good! Have not tested it but I assume you did.

**Review by Mic92 — APPROVED** [review]
Looks good! Have not tested it but I assume you did.

**Review by Mic92 — APPROVED** [review]
Looks good! Have not tested it but I assume you did.

**Pull request merged**